## Pandas in Python

### Reading a CSV file using Pandas

Pandas is typically used for working in tabular data (simliar to the data stored in a spreadsheet). Pandas provides helper functions to read data from various file formates like CSV, Excel spreadsheets, HTML tables, JSON, SQL and more. Let's download a file `italy-covid-daywise.txt` which contains daywise Covid-19 data for Italy in the following format:

```
date,new_cases,new_deaths,new_tests
2020-04-21,2256.0,454.0,28095.0
2020-04-22,2729.0,534.0,44248.0
2020-04-23,3370.0,437.0,37083.0
2020-04-24,2646.0,464.0,95273.0
2020-04-25,3021.0,420.0,38676.0
2020-04-26,2357.0,415.0,24113.0
2020-04-27,2324.0,260.0,26678.0
2020-04-28,1739.0,333.0,37554.0
...
```

This format of storing data is known as *comma separated values* or CSV. 

> **CSVs**: A comma-separated values (CSV) file is a delimited text file that uses a comma to separate values. Each line of the file is a data record. Each record consists of one or more fields, separated by commas. A CSV file typically stores tabular data (numbers and text) in plain text, in which case each line will have the same number of fields. (Wikipedia)


We'll download this file using the `urlretrieve` function from the `urllib.request` module.

To read the file, we can use the `read_csv` method from Pandas. Let's being by importing the Pandas library. It is typically imported with the alias `pd`.

In [1]:
import pandas as pd

In [2]:
covid_df = pd.read_csv("italy-covid-daywise.csv")

In [3]:
type(covid_df)

pandas.DataFrame

In [4]:
covid_df.dtypes

date              str
new_cases       int64
new_deaths      int64
new_tests     float64
dtype: object

In [5]:
covid_df

,date,new_cases,new_deaths,new_tests
0,31-12-2019,0,0,NaN
1,01-01-2020,0,0,NaN
2,02-01-2020,0,0,NaN
3,03-01-2020,0,0,NaN
4,04-01-2020,0,0,NaN
...,...,...,...,...
243,30-08-2020,1444,1,53541.0
244,31-08-2020,1365,4,42583.0
245,01-09-2020,996,6,54395.0
246,02-09-2020,975,8,NaN


We can view some basic information about the data frame using the `.info` method.

In [6]:
covid_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        248 non-null    str    
 1   new_cases   248 non-null    int64  
 2   new_deaths  248 non-null    int64  
 3   new_tests   135 non-null    float64
dtypes: float64(1), int64(2), str(1)
memory usage: 7.9 KB


In [7]:
covid_df.describe()

,new_cases,new_deaths,new_tests
count,248.000000,248.000000,135.000000
mean,1094.818548,143.133065,31699.674074
std,1554.508002,227.105538,11622.209757
min,-148.000000,-31.000000,7841.000000
25%,123.000000,3.000000,25259.000000
50%,342.000000,17.000000,29545.000000
75%,1371.750000,175.250000,37711.000000
max,6557.000000,971.000000,95273.000000


The `columns` property contains the list of columns within the data frame.

In [8]:
covid_df.columns

Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='str')

You can also retrieve the number of rows and columns in the data frame using the `.shape` method

In [9]:
covid_df.shape

(248, 4)

Here's a summary of the functions & methods we've looked at so far:

* `pd.read_csv` - Read data from a CSV file into a Pandas `DataFrame` object
* `.info()` - View basic infomation about rows, columns & data types
* `.describe()` - View statistical information about numeric columns
* `.columns` - Get the list of column names
* `.shape` - Get the number of rows & columns as a tuple

## Retrieving data from a data frame

The first thing you might want to do is to retrieve data from this data frame e.g. the counts of a specific day or the 
list of values in a specific column. To do this, it might help to understand the internal representation of data in a data frame. Conceptually, you can think of a dataframe as a dictionary of lists: the keys are column names, and the values are lists/arrays containing data for the respective columns. 

In [10]:
# Pandas format is simliar to this
covid_data_dict = {
    'date':       ['2020-08-30', '2020-08-31', '2020-09-01', '2020-09-02', '2020-09-03'],
    'new_cases':  [1444, 1365, 996, 975, 1326],
    'new_deaths': [1, 4, 6, 8, 6],
    'new_tests': [53541, 42583, 54395, None, None]
}

Representing data in the above format has a few benefits:

* All values in a column typically have the same type of value, so it's more efficient to store them in a single array.
* Retrieving the values for a particular row simply requires extracting the elements at a given index from each of the column arrays.
* The representation is more compact (column names are recorded only once) compared to other formats where you might use a dictionary for each row of data (example shown below).

In [11]:
# Pandas format is not similar to this
covid_data_list = [
    {'date': '2020-08-30', 'new_cases': 1444, 'new_deaths': 1, 'new_tests': 53541},
    {'date': '2020-08-31', 'new_cases': 1365, 'new_deaths': 4, 'new_tests': 42583},
    {'date': '2020-09-01', 'new_cases': 996, 'new_deaths': 6, 'new_tests': 54395},
    {'date': '2020-09-02', 'new_cases': 975, 'new_deaths': 8 },
    {'date': '2020-09-03', 'new_cases': 1326, 'new_deaths': 6},
]

With the dictionary of lists analogy in mind, we can now guess how we might be able to retrieve data from a data frame. for example, we can get a list of values from a specific column using the `[]` indexing notation.

In [12]:
covid_df['new_cases']

0         0
1         0
2         0
3         0
4         0
       ... 
243    1444
244    1365
245     996
246     975
247    1326
Name: new_cases, Length: 248, dtype: int64

In [13]:
covid_data_dict['new_cases']

[1444, 1365, 996, 975, 1326]

In [14]:
covid_data_list['new_cases']

TypeError: list indices must be integers or slices, not str

In [15]:
type(covid_df['new_cases'])

pandas.Series

In [16]:
type(covid_data_dict['new_cases'])

list

Just like arrays, you can retrieve a specific value with a series using the indexing notation `[]`.

In [17]:
print(covid_df['new_cases'][234])

840


In [18]:
print(covid_df['new_cases'][246])

975


Pandas also provides the `.at` method to directly retrieve at a specific row & column.

In [19]:
covid_df.at[246, 'new_cases']

np.int64(975)

Instead of using the indexing notation `[]`, Pandas also allows accessing columns as properties of the data frame using the `.` notation. However, this method only works for columns whose names do not contain spaces or special chracters.

In [20]:
covid_df.new_cases

0         0
1         0
2         0
3         0
4         0
       ... 
243    1444
244    1365
245     996
246     975
247    1326
Name: new_cases, Length: 248, dtype: int64

Further, you can also pass a list of columns within the indexing notation `[]` to access a subset of the data frame with just the given columns.

In [21]:
cases_df = covid_df[['date', 'new_cases']]
cases_df

,date,new_cases
0,31-12-2019,0
1,01-01-2020,0
2,02-01-2020,0
3,03-01-2020,0
4,04-01-2020,0
...,...,...
243,30-08-2020,1444
244,31-08-2020,1365
245,01-09-2020,996
246,02-09-2020,975


Note, however, that the new data frame `cases_df` is simply a "view" of the original data frame `covid_df` i.e. they both point to the same data in the computer's memory, and changing any values inside one of them will also change the respective values in the other. Sharing data between data frames makes data manipulation in Pandas blazing fast, and you needn't worry about the overhead of copying thousands or millions of rows every time you want to crate a new data frame by operating on an exissting one.

Sometimes you might need a full copy of the data frame, in which case you can use the `copy` method.

In [22]:
covid_df_copy = covid_df.copy()

The data within `covid_df_copy` is completely separate from `covid_df`, and changing values inside one of them will not affect the other.

To access a specific row of data Pandas provides the `.loc` method.

In [23]:
covid_df

,date,new_cases,new_deaths,new_tests
0,31-12-2019,0,0,NaN
1,01-01-2020,0,0,NaN
2,02-01-2020,0,0,NaN
3,03-01-2020,0,0,NaN
4,04-01-2020,0,0,NaN
...,...,...,...,...
243,30-08-2020,1444,1,53541.0
244,31-08-2020,1365,4,42583.0
245,01-09-2020,996,6,54395.0
246,02-09-2020,975,8,NaN


In [24]:
covid_df.loc[243]

date          30-08-2020
new_cases           1444
new_deaths             1
new_tests        53541.0
Name: 243, dtype: object

Each retrieved row is also a `Series` object.

In [25]:
type(covid_df.loc[243])

pandas.Series

To view the first or last few rows of data, we can use the `.head` and `.tail` methods.

In [26]:
covid_df.head(5)

,date,new_cases,new_deaths,new_tests
0,31-12-2019,0,0,NaN
1,01-01-2020,0,0,NaN
2,02-01-2020,0,0,NaN
3,03-01-2020,0,0,NaN
4,04-01-2020,0,0,NaN


In [27]:
covid_df.tail(5)

,date,new_cases,new_deaths,new_tests
243,30-08-2020,1444,1,53541.0
244,31-08-2020,1365,4,42583.0
245,01-09-2020,996,6,54395.0
246,02-09-2020,975,8,NaN
247,03-09-2020,1326,6,NaN


Notice above that while the first few values in the `new_cases` and `new_deaths` columns are `0`, the corresponding values within the `new_tests` column are `NaN`. That is because the CSV file does not contain any data for the `new_tests` column for certain dates (you can verify this by looking into the file). It's possible that these values are missing or unknown.

In [28]:
covid_df.at[0, 'new_tests']

np.float64(nan)

In [29]:
type(covid_df.at[0, 'new_tests'])

numpy.float64

The distinction between `0` and `NaN` is subtle but important. In this dataset, it represents that daily test numbers were not reported on specific dates. In fact, Italy started reporting daily tests on April 19, 2020. By that time, 935310 tests had already been conducted. 

We can find the first index that doesn't contain a `NaN` value using `first_valid_index` method of a series.

In [61]:
covid_df.new_tests.first_valid_index()

111

Let's look at a few rows before and after this index to verify that the values indeed change from `NaN` to actual numbers. We can do this by passing a range to `loc`.

In [31]:
covid_df.loc[108:113]

,date,new_cases,new_deaths,new_tests
108,17-04-2020,3786,525,NaN
109,18-04-2020,3493,575,NaN
110,19-04-2020,3491,480,NaN
111,20-04-2020,3047,433,7841.0
112,21-04-2020,2256,454,28095.0
113,22-04-2020,2729,534,44248.0


The `.sample` method can be used to retrieve a random sample of rows from the data frame.

In [32]:
covid_df.sample(10)

,date,new_cases,new_deaths,new_tests
99,08-04-2020,3039,604,NaN
165,13-06-2020,163,56,32880.0
140,19-05-2020,451,99,40226.0
92,01-04-2020,4053,839,NaN
75,15-03-2020,3497,173,NaN
100,09-04-2020,3836,540,NaN
216,03-08-2020,238,8,13467.0
68,08-03-2020,1247,36,NaN
232,19-08-2020,401,5,41290.0
149,28-05-2020,584,117,39838.0


Notice that even though we have taken a random sample, the original index of each row has been preserved. This is an important an useful property of data frames - each row of data has an index associated with it.

Here's a summary of the functions & methods we looked at in this section:

- `covid_df['new_cases']` - retrieving columns as series using a column name
- `new_cases[243]` - retrieving values from a series using an index
- `covid_df.at[243, 'new_cases']` - retrieving a single value from a data frame
- `covid_df.copy()` - Creating a deep copy of a data frame
- `covid_df.loc[243]` - Retrieving a row or range of rows of data from the data frame
- `head`, `tail` and `sample` - Retrieving multiple rows of data from the data frame
- `covid_df.new_tests.first_valid_index` - Finding the first non-empty index in a series

## Analyzing Data from data frames

Let's try to answer some questions about our data.

**Q: What is the total number of reported cases and deaths related to Covid-19 in Italy?**

Similar to Numpy arrays, a Pandas series supports the `sum` method to answer these questions.

In [33]:
total_cases = covid_df.new_cases.sum()

In [34]:
total_deaths = covid_df.new_deaths.sum()

We can check the first non-NaN index using `first_valid_index`

In [35]:
print('The number of reported cases is {} and the number of reported deaths is {}.'.format(int(total_cases), int(total_deaths)))

The number of reported cases is 271515 and the number of reported deaths is 35497.


**Q: What is the overall death rate (ratio of reported deats to reported cases)?**

In [36]:
death_rate = covid_df.new_deaths.sum() / covid_df.new_cases.sum()

In [37]:
print("The overall reported death rate in Italy is {:.2f} %.".format(death_rate*100))

The overall reported death rate in Italy is 13.07 %.


**Q: What is the overall number of tests conducted? A total of 935310 tests were conducted before daily test numbers were being reported.**

In [38]:
initial_tests = 935310
total_tests = initial_tests + covid_df.new_tests.sum()

In [39]:
print(total_tests)

5214766.0


**Q: What fraction of test returned a postive result?**

In [40]:
positive_rate = total_cases / total_tests

In [41]:
print('{:.2f}% of tests in Italy led to a positive diagnosis.'.format(positive_rate*100))

5.21% of tests in Italy led to a positive diagnosis.


Try asking and answering some more questions about the data using the empty cells below.

## Querying and sorting rows

Let's say we want only want to look at the days which had more than 1000 reported cases. We can use a boolean expression to chech which rows satisfy this criterion.

In [42]:
high_new_cases  = covid_df.new_cases > 1000

In [43]:
high_new_cases

0      False
1      False
2      False
3      False
4      False
       ...  
243     True
244     True
245    False
246    False
247     True
Name: new_cases, Length: 248, dtype: bool

The boolean expression returns a series containing `True` and `False` boolean values. This series can be used to filter out just those rows where the value in the series is `True`. The result is data frame with a subset of rows from the original.

In [44]:
covid_df[high_new_cases]

,date,new_cases,new_deaths,new_tests
68,08-03-2020,1247,36,NaN
69,09-03-2020,1492,133,NaN
70,10-03-2020,1797,98,NaN
72,12-03-2020,2313,196,NaN
73,13-03-2020,2651,189,NaN
...,...,...,...,...
241,28-08-2020,1409,5,65135.0
242,29-08-2020,1460,9,64294.0
243,30-08-2020,1444,1,53541.0
244,31-08-2020,1365,4,42583.0


We can write this succintly on a single line by passing the boolean expression as an index to the data frame.

In [45]:
high_cases_df = covid_df[covid_df.new_cases > 1000]

In [46]:
high_cases_df

,date,new_cases,new_deaths,new_tests
68,08-03-2020,1247,36,NaN
69,09-03-2020,1492,133,NaN
70,10-03-2020,1797,98,NaN
72,12-03-2020,2313,196,NaN
73,13-03-2020,2651,189,NaN
...,...,...,...,...
241,28-08-2020,1409,5,65135.0
242,29-08-2020,1460,9,64294.0
243,30-08-2020,1444,1,53541.0
244,31-08-2020,1365,4,42583.0


The data frame contains 72 rows, but only the first 5 & last 5 rows are displayed by default with Jupyter, for brevity. To view, all the rows, we can modify some display options.

In [47]:
from IPython.display import display
with pd.option_context('display.max_rows', 100): 
    display(covid_df[covid_df.new_cases > 1000])

,date,new_cases,new_deaths,new_tests
68,08-03-2020,1247,36,NaN
69,09-03-2020,1492,133,NaN
70,10-03-2020,1797,98,NaN
72,12-03-2020,2313,196,NaN
73,13-03-2020,2651,189,NaN
74,14-03-2020,2547,252,NaN
75,15-03-2020,3497,173,NaN
76,16-03-2020,2823,370,NaN
77,17-03-2020,4000,347,NaN
78,18-03-2020,3526,347,NaN


We can also formulate more complex queries that involve mutiple columns. As an example, let's try to determine the days when the ratio of cases reported to tests conducted is higher than the overall `positive_rate`.

In [48]:
print(positive_rate)

0.05206657403227681


In [49]:
high_ratio_df = covid_df[covid_df.new_cases / covid_df.new_tests > positive_rate]

In [50]:
high_cases_df

,date,new_cases,new_deaths,new_tests
68,08-03-2020,1247,36,NaN
69,09-03-2020,1492,133,NaN
70,10-03-2020,1797,98,NaN
72,12-03-2020,2313,196,NaN
73,13-03-2020,2651,189,NaN
...,...,...,...,...
241,28-08-2020,1409,5,65135.0
242,29-08-2020,1460,9,64294.0
243,30-08-2020,1444,1,53541.0
244,31-08-2020,1365,4,42583.0


Performing operations on multiple columns results in a new series.

In [51]:
covid_df.new_cases / covid_df.new_tests

0           NaN
1           NaN
2           NaN
3           NaN
4           NaN
         ...   
243    0.026970
244    0.032055
245    0.018311
246         NaN
247         NaN
Length: 248, dtype: float64

In [52]:
covid_df['positive_rate'] = covid_df.new_cases / covid_df.new_tests

In [53]:
covid_df

,date,new_cases,new_deaths,new_tests,positive_rate
0,31-12-2019,0,0,NaN,NaN
1,01-01-2020,0,0,NaN,NaN
2,02-01-2020,0,0,NaN,NaN
3,03-01-2020,0,0,NaN,NaN
4,04-01-2020,0,0,NaN,NaN
...,...,...,...,...,...
243,30-08-2020,1444,1,53541.0,0.026970
244,31-08-2020,1365,4,42583.0,0.032055
245,01-09-2020,996,6,54395.0,0.018311
246,02-09-2020,975,8,NaN,NaN


Keep in mind however, sometimes it takes a few days to get the results for a test, so we can't really compare the number of new cases with the number of tests conducted on the same day. Any inference based on this `positive_rate` column is likely to be incorrect. It's important to watch out for subtle relationships like these which are often not conveyed within the CSV file and require some external context. It's always a good idea to read through the documentation provided with the dataset or ask for more information.

For now, let's remove the `positive_rate` column using the `drop` method.

In [54]:
covid_df.drop(columns = ['positive_rate'], inplace = True)

### Sorting rows using column values

The rows can also be sorted by a specific column using `.sort_values`. Let's sort to identify the days with the highest number of cases, then chain it with the `head` method to get the 10 days with the most cases.

In [55]:
covid_df.sort_values('new_cases', ascending = False).head(10)

,date,new_cases,new_deaths,new_tests
82,22-03-2020,6557,795,NaN
87,27-03-2020,6153,660,NaN
81,21-03-2020,5986,625,NaN
89,29-03-2020,5974,887,NaN
88,28-03-2020,5959,971,NaN
83,23-03-2020,5560,649,NaN
80,20-03-2020,5322,429,NaN
85,25-03-2020,5249,743,NaN
90,30-03-2020,5217,758,NaN
86,26-03-2020,5210,685,NaN


It looks like the last two weeks of March had the highest number of daily cases. Let's compare this to the days where the highest number of deaths were recorded.

In [56]:
covid_df.sort_values('new_deaths', ascending = False).head(10)

,date,new_cases,new_deaths,new_tests
88,28-03-2020,5959,971,NaN
89,29-03-2020,5974,887,NaN
92,01-04-2020,4053,839,NaN
91,31-03-2020,4050,810,NaN
82,22-03-2020,6557,795,NaN
95,04-04-2020,4585,764,NaN
94,03-04-2020,4668,760,NaN
90,30-03-2020,5217,758,NaN
85,25-03-2020,5249,743,NaN
93,02-04-2020,4782,727,NaN


It appears that daily deaths hit a peak just about week after the peak in the daily new cases.

Let's also look at the days with the least number of cases. We might expect to see the first few days of the year in this list.

In [57]:
covid_df.sort_values('new_cases').head(10)

,date,new_cases,new_deaths,new_tests
172,20-06-2020,-148,47,29875.0
0,31-12-2019,0,0,NaN
2,02-01-2020,0,0,NaN
1,01-01-2020,0,0,NaN
4,04-01-2020,0,0,NaN
5,05-01-2020,0,0,NaN
6,06-01-2020,0,0,NaN
3,03-01-2020,0,0,NaN
8,08-01-2020,0,0,NaN
9,09-01-2020,0,0,NaN


Seems like the count of new cases on June 20th was `-148`, a negative number! Not something we might have expected, but that's the nature of real world data. It could simply be a data entry error, or it's possible that the government may have issued a correction to account for miscounting in the past. Can you dig through news articles online and figure out why the number was negative?

Let's look at the some of days before and after June 20th.

In [58]:
covid_df.loc[169:175]

,date,new_cases,new_deaths,new_tests
169,17-06-2020,210,34,33957.0
170,18-06-2020,328,43,32921.0
171,19-06-2020,331,66,28570.0
172,20-06-2020,-148,47,29875.0
173,21-06-2020,264,49,24581.0
174,22-06-2020,224,24,16152.0
175,23-06-2020,221,23,23225.0


If this was indeed a data entry error, we can use one of the following approaches for dealing with the missing or faulty value:
1. Replace it with `0`.
2. Repalce it with the average of the entire column
3. Replace it with the average of the values on the previous & next date
4. Discard the row entirely

Which approach you pick requires some context about the data and the problem. In this case, since we are dealing data ordered by date, we can pick approach 3. 

The `.at` method can be used to modify a specific value within the data frame.

In [62]:
numeric_cols = covid_df.select_dtypes(exclude = 'str').columns

In [63]:
covid_df[numeric_cols] = covid_df[numeric_cols].astype(float)

In [64]:
covid_df.at[172, 'new_cases'] = (covid_df.at[171, 'new_cases'] + covid_df.at[172, 'new_cases'])/2

<class 'pandas.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        248 non-null    str    
 1   new_cases   248 non-null    float64
 2   new_deaths  248 non-null    float64
 3   new_tests   135 non-null    float64
dtypes: float64(3), str(1)
memory usage: 7.9 KB
